### ISSUE TITLE
Config clarification to benefit from Listwise losses and scoring during model training & inference

### ISSUE THREAD

**Issue by gever-ashdodi** [tensorflow/ranking#267]
Hello TensorFlow Ranking Team,

I am looking for some help in understanding some of the Framework's configuration parameters and to solidify my own understanding about the behavior of the Framework.

#### TL;DR 
I would like to train a model to benefit from Listwise scoring & losses during training AND Listwise scoring during inference.

#### Details
I went through the [Handling Sparse Features Collab tutorial](https://colab.research.google.com/github/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb) AND following issues: https://github.com/tensorflow/ranking/issues/83, https://github.com/tensorflow/ranking/issues/132, https://github.com/tensorflow/ranking/issues/201 and https://github.com/tensorflow/ranking/issues/235 and the following are my conclusions about the Framework configuration and the resulting behavior:

1. The loss function must be set to `tfr.losses.RankingLossKey.APPROX_NDCG_LOSS` to benefit from the approximation for the NDCG metric. 

The group_size parameter (i.e.: `_GROUP_SIZE`). This is where things start getting a little vague for me. 

2. According to the aforementioned Collab tutorial comment, when `_GROUP_SIZE = 1` each document is scored independently during training & inference (i.e.: `Pointwise` scoring). But, according to https://github.com/tensorflow/ranking/issues/201#issuecomment-637274452 (@xuanhuiwang's comment) even with `_GROUP_SIZE = 1`, the model benefits from Listwise _losses_ during training. 
I am wondering - why/how is it possible to still benefit from Listwise losses during training when the _scoring_ is Pointwise? Is it because `(a)` I set the loss function to `tfr.losses.RankingLossKey.APPROX_NDCG_LOSS` or `(b)` is it because as an input to the model we pass in ELWCs, where each ELWC contains a list of Examples (i.e.: a whole list is an input)?

3. If I want to benefit from Listwise _scoring_ during training, then the  `_GROUP_SIZE` must be > 2, But, if I set the group size to be greater than 2, some variance in predictions will be observed during inference time (due to random shuffle), according to https://github.com/tensorflow/ranking/issues/201#issuecomment-637274452 and #235.
In other words, when trained with `_GROUP_SIZE > 2`, the model _cannot_ benefit from Listwise _scoring_ during inference in a deterministic way, thus I must train my model with `_GROUP_SIZE = 1` (Pointwise _scoring_). Is this correct? 

4. If I do want to benefit from Listwise _scoring_ during _inference_, then I should look into adopting the approach to capture cross-document interactions during training (e.g.: [tensorflow_ranking/examples/tf_ranking_tfrecord.py#L229](https://github.com/tensorflow/ranking/blob/4084728fe7dae65c76474025a5d2c4ff9b7cf73a/tensorflow_ranking/examples/tf_ranking_tfrecord.py#L229)), while keeping the `_GROUP_SIZE = 1`. Is this correct? 

I appreciate your patience and understanding!

cc: @HongleiZhuang @xuanhuiwang

**Comment by gever-ashdodi** [comment]
Hello Team,

I do realize that some of these questions may sound naive or trivial, but could you please advise?

**Comment by xuanhuiwang** [comment]
Please see my replies blow.

2
The input is ELWC which has a list of examples. So after scoring, we will have all the scores of a single ELWC in the same batch, thus the pairwise and listwise loss can be computed for all examples in a single ELWC. Please note that the logits and labels have shape [batch_size, list_size]. 

2 (a)
There are many other ranking losses beside tfr.losses.RankingLossKey.APPROX_NDCG_LOSS, such as Pairwise logitistic loss and softmax loss.

2(b)
See my replies above on how listwise losses can be used with pointwise scoring.

3
Yes.

4
A keras version of the DIN is here: https://github.com/tensorflow/ranking/blob/4084728fe7dae65c76474025a5d2c4ff9b7cf73a/tensorflow_ranking/examples/keras/antique_kpl_din.py. This is the currently recommended way to model document interaction in TF-Ranking.

**Comment by gever-ashdodi** [comment]
Thank you so much, @xuanhuiwang

**Comment by xuanhuiwang** [comment]
I am going to close this issue. Feel free to reopen if there is more questions.
